python/shotgun_model/data_handler.py

```diff
@@ -0,0 +1,473 @@
+            # nothing to do
+            return
+
+        self._log_debug("Unloading in-memory cache for %s" % self)
+        self._cache = None
+
+    @log_timing
+    def save_cache(self):
+        """
+        Saves the current cache to disk.
+        """
+        self._log_debug("Saving to disk: %s" % self)
+
+        # try to create the cache folder with as open permissions as possible
+        cache_dir = os.path.dirname(self._cache_path)
+
+        # make sure the cache directory exists
+        # todo: upgrade to 0.18 filesystem methods
+        if not os.path.exists(cache_dir):
+            try:
+                os.makedirs(cache_dir, 0777)
+            except OSError, e:
+                # Race conditions are perfectly possible on some network
+                # storage setups so make sure that we ignore any file
+                # already exists errors, as they are not really errors!
+                if e.errno != errno.EEXIST:
+                    # re-raise
+                    raise
+
+        # now write the file
+        old_umask = os.umask(0)
+        try:
+            with open(self._cache_path, "wb") as fh:
+                pickler = cPickle.Pickler(fh, protocol=2)
+                # speeds up pickling but only works when there
+                # are no cycles in the data set
+                # pickler.fast = 1
+
+                # TODO: we are currently storing a parent node in our data structure
+                # for performance and cache size. By removing this, we could turn
+                # on the fast mode and this would speed things up further.
+
+                pickler.dump(self.FORMAT_VERSION)
+                pickler.dump(self._cache)
+
+            # and ensure the cache file has got open permissions
+            os.chmod(self._cache_path, 0666)
+
+        finally:
+            # set mask back to previous value
+            os.umask(old_umask)
+
+        self._log_debug("Completed save of %s. Size %s bytes" % (self, os.path.getsize(self._cache_path)))
+
+    def get_data_item_from_uid(self, unique_id):
+        """
+        Given a unique id, return a :class:`ShotgunDataItem`
+        Returns None if the given uid is not present in the cache.
+
+        :param unique_id: unique identifier, typically an int or a string
+        :returns: :class:`ShotgunDataItem`
+        """
+        # avoid cyclic imports
+        from .data_item import ShotgunDataItem
+
+        if not self.is_cache_loaded():
+            return None
+
+        return ShotgunDataItem(self._cache[self.CACHE_BY_UID].get(unique_id))
+
+    @log_timing
+    def generate_child_nodes(self, unique_id, parent_object, factory_fn):
+        """
+        Generate nodes recursively from the data set
+
+        each node will be passed to the factory method for construction.
+
+        unique id can be none, meaning generate the top level of the tree
+
+        :param unique_id:     Unique identifier, typically an int or a string
```

Comment:
Why is it strings or ints depending on the occasion? What does that imply? What are the consequences of that? Maybe this could be explained.